### PR TITLE
WPB-4835 call-between-web-and-android-cant-be-established-when-one-backend-is-not-available

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-4835

Implementation according to https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/897712162/Use+Case+Fetching+the+clients+of+a+user

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: integration/test/Test/Client.hs [hunk 1]
```diff
@@ -32,3 +41,44 @@ testClientLastActive = do
...
+          [ API.createFedConn x (API.FedConn y "full_search")
+            | x <- domains,
+              y <- domains,
+              x /= y
+          ]
+
+      downUser <- randomUser downBackend.berDomain def
+      connectUsers ownUser1 downUser
+      connectUsers ownUser2 downUser
+      pure (downUser)
```
Comment (akshaymankar) [inline integration/test/Test/Client.hs:80]: This user doesn't even need to exist if the backend is down anyway. Might save time in actually having to start the backend.

File: integration/test/Test/Client.hs [hunk 2]
```diff
@@ -32,3 +41,44 @@ testClientLastActive = do
...
+
+    bindResponse (listUsersClients ownUser1 [ownUser1, ownUser2, downUser]) $ \resp -> do
+      resp.status `shouldMatchInt` 200
+      resp.json %. "qualified_user_map" `shouldMatch` expectedResponse
```
Comment (akshaymankar) [inline integration/test/Test/Client.hs:84]: Is this OK for old clients? As in what do they do if they don't see some users? Do they have code to handle this? If not we should consider adding an extra field here which indicates partial failure and doing this in the next API version.

File: services/brig/src/Brig/API/Client.hs [hunk 3]
```diff
@@ -133,13 +132,18 @@ lookupPubClientsBulk :: [Qualified UserId] -> ExceptT ClientError (AppT r) (Qual
...
   pure $ QualifiedUserMap (Map.union localUserClientMap remoteUserClientMap)
+  where
+    getRemoteClients :: Map Domain [UserId] -> AppT r (Map Domain (UserMap (Set PubClient)))
+    getRemoteClients uids = do
+      results <-
+        traverse
+          ( \(d, ids) -> fmap (d,) <$> runExceptT (getUserClients d (GetUserClients ids))
+          )
+          (Map.toList uids)
+      pure $ Map.fromList (rights results)
```
Comment (akshaymankar) [inline services/brig/src/Brig/API/Client.hs:146]: It maybe worth logging the errors?

File: integration/test/Test/Client.hs (hunk 2 above)
Comment (marcoconti83) [reply integration/test/Test/Client.hs:84]: Before this PR, the request will return 503. After the PR, the request will return a subset of the users.
No matter how the clients are able to handle a subset of the users (and they seem to be doing it properly already), the 503 was causing a DOS anyway that would prevent calling from working completely, so we are not making things worse :)

Comment (akshaymankar) [reply integration/test/Test/Client.hs:84]: Alright, as far as clients don't think the users are deleted :smile:

**Review by akshaymankar — APPROVED** [review]

File: integration/test/Test/Client.hs (hunk 2 above)
Comment (akshaymankar) [reply integration/test/Test/Client.hs:84]: Maybe we should add this information to the description of the endpoint tho.

File: libs/wire-api/src/Wire/API/Routes/Public/Brig.hs [hunk 4]
```diff
@@ -920,6 +920,7 @@ type ClientAPI =
     :<|> Named
            "list-clients-bulk@v2"
            ( Summary "List all clients for a set of user ids"
+               :> Description "If a backend is unreachable, the clients from this backend will be omitted from the response"
```
Comment (akshaymankar) [inline libs/wire-api/src/Wire/API/Routes/Public/Brig.hs:923]:
```suggestion
               :> Description "If a backend is unreachable, the clients from that backend will be omitted from the response"
```

**Review by akshaymankar — APPROVED** [review]

**Pull request merged**